tests/e2e/vmlistener_test.go

```diff
@@ -0,0 +1,123 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This test suite holds testcases related with vm listener functionality.
+// In case of vm/esx host restart vDVS cleans up stale attachment details
+// the volume.
+
+package e2e
+
+import (
+	"log"
+	"os"
+	"time"
+
+	. "gopkg.in/check.v1"
+
+	"github.com/vmware/docker-volume-vsphere/tests/utils/admincli"
+	"github.com/vmware/docker-volume-vsphere/tests/utils/dockercli"
+	"github.com/vmware/docker-volume-vsphere/tests/utils/esxcli"
+	"github.com/vmware/docker-volume-vsphere/tests/utils/govc"
+	"github.com/vmware/docker-volume-vsphere/tests/utils/inputparams"
+	"github.com/vmware/docker-volume-vsphere/tests/utils/misc"
+	"github.com/vmware/docker-volume-vsphere/tests/utils/verification"
+)
+
+type VMListenerTestParams struct {
+	volumeName     string
+	dockerHostIP   string
+	dockerHostName string
+	containerName  string
+	esxIP          string
+}
+
+func (s *VMListenerTestParams) SetUpSuite(c *C) {
+	s.dockerHostIP = os.Getenv("VM2")
+	s.dockerHostName = govc.RetrieveVMNameFromIP(s.dockerHostIP)
+	s.esxIP = os.Getenv("ESX")
+}
+
+func (s *VMListenerTestParams) SetUpTest(c *C) {
+	s.volumeName = inputparams.GetVolumeNameWithTimeStamp("vmListenerTest")
+	s.containerName = inputparams.GetContainerNameWithTimeStamp("vmListenerTest")
+}
+
+func (s *VMListenerTestParams) TearDownTest(c *C) {
+	// Note: no need to remove container, it is already removed as
+	// vm got killed and `--restart=always` flag is not added
+	out, err := dockercli.DeleteVolume(s.dockerHostIP, s.volumeName)
+	c.Assert(err, IsNil, Commentf(misc.FormatOutput(out)))
+}
+
+var _ = Suite(&VMListenerTestParams{})
+
+// Test vmdkops service restart + kill vm process
+// 1. Create a volume
+// 2. Attach it to a container
+// 3. Restart the vmdkops service
+// 4. Verify vmdkops_admin volume ls > verify volume stays as attached
+// 5. Kill the vm
+// 6. Verification: volume status should be detached
+
+func (s *VMListenerTestParams) TestKillVM(c *C) {
+	log.Printf("START: Test vmdkops service restart + kill vm process")
+
+	out, err := dockercli.CreateVolume(s.dockerHostIP, s.volumeName)
+	c.Assert(err, IsNil, Commentf(misc.FormatOutput(out)))
+
+	out, err = dockercli.AttachVolume(s.dockerHostIP, s.volumeName, s.containerName)
+	c.Assert(err, IsNil, Commentf(misc.FormatOutput(out)))
+
+	status := verification.VerifyAttachedStatus(s.volumeName, s.dockerHostIP, s.esxIP)
+	c.Assert(status, Equals, true, Commentf("Volume %s is not attached", s.volumeName))
+
+	// restart vmdkops service
+	out, err = admincli.RestartVmdkopsService(s.esxIP)
+	c.Assert(err, IsNil, Commentf(misc.FormatOutput(out)))
+
+	// make sure volume stays attached
+	status = verification.VerifyAttachedStatus(s.volumeName, s.dockerHostIP, s.esxIP)
```

Comment:
Restarting vmdkosp service won't change the status of volume here. Any other reason you have to do this check  again after check on L82? Nothing bad, just analyzing the necessity, given each check involves remote ssh and is costly operation in my view.